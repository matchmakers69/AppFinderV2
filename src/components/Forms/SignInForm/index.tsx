import { FC, useState } from "react";
import { yupResolver } from "@hookform/resolvers/yup/dist/yup";
import Cookies from "js-cookie";
import { useForm } from "react-hook-form";
import { useDispatch } from "react-redux";
import * as Yup from "yup";
import loginSchema from "src/lib/validation/loginFormSchema";
import { logInAction } from "src/store/auth/actions";

export type SignInType = Pick<Yup.InferType<typeof loginSchema>, "username" | "password">;

type SignInFormProps = {
  login: (user: SignInType) => Promise<SignInType>;
};

const SignInForm: FC<SignInFormProps> = ({ login }) => {
  const [loading, setLoading] = useState(false);
  const dispatch = useDispatch();
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isValid },
  } = useForm<SignInType>({
    mode: "onChange",
    resolver: yupResolver(loginSchema),
  });
  const onLoginFormSubmit = async (data: SignInType): Promise<void> => {
    setTimeout(() => {
      reset({
        username: "",
        password: "",
      });
    }, 300);
    setLoading(true);
    try {
      const response = await login(data);
      dispatch(logInAction(response));
      // set cookies
      Cookies.set("token", "here jwt token");
      setLoading(false);
    } catch (err) {
      console.log(err);
      setLoading(false);
    }
  };
  const isDisabled = !isValid || loading;
  if (loading) {
    return <span>Is loading...</span>;
  }
  return (
    <form onSubmit={handleSubmit(onLoginFormSubmit)} noValidate>
      <div>
        <label>
          Username
          <input
            type="text"
            {...register("username")}
            aria-invalid={!!errors.username}
            autoComplete="username"
            aria-describedby="password-constraints"
            required
          />
        </label>
        {errors.username ? <span role="alert">{errors.username.message}</span> : null}
      </div>
      <div>
        <label>
          Password
          <input
            type="password"
            {...register("password")}
            aria-invalid={!!errors.password}
            aria-describedby="password-constraints"
            autoComplete="current-password"
            required
          />
        </label>

        {errors.password ? <span role="alert">{errors.password.message}</span> : null}
      </div>
      <button disabled={isDisabled} type="submit">
        Submit
      </button>
    </form>
  );
};

export default SignInForm;
