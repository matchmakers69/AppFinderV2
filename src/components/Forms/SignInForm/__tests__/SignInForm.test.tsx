import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { errorMessages } from "src/utils/errors/errorMessages";
import SignInForm, { SignInType } from "..";

const signInMock = jest.fn((user: SignInType) => {
  return Promise.resolve(user);
});

const mockStore = configureStore();
afterEach(() => jest.resetAllMocks());

describe("<SignInForm />", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  it("should create a snapshot for SignInForm", () => {
    const initialState = {
      isLoggedIn: false,
      user: null,
      loading: false,
    };
    const store = mockStore(initialState);
    expect(
      renderer
        .create(
          <Provider store={store}>
            <SignInForm login={signInMock} />
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("displays matching errors when the values are invalid", async () => {
    const initialState = {
      isLoggedIn: false,
      user: null,
      loading: false,
    };
    const store = mockStore(initialState);
    render(
      <Provider store={store}>
        <SignInForm login={signInMock} />
      </Provider>
    );

    const submitButton = screen.getByRole("button", { name: /Submit/ });
    const usernameInput = screen.getByLabelText("Username");
    const passwordInput = screen.getByLabelText("Password");
    const username = "test";
    const password = "test";

    userEvent.type(usernameInput, username);
    userEvent.type(passwordInput, password);
    userEvent.click(submitButton);

    const usernamePatternErrorMessage = await screen.findByText(errorMessages.username.pattern);
    const passwordPatternErrorMessage = await screen.findByText(errorMessages.password.pattern);

    expect(signInMock).not.toBeCalled();
    expect(usernamePatternErrorMessage).toBeInTheDocument();
    expect(passwordPatternErrorMessage).toBeInTheDocument();
  });
});
