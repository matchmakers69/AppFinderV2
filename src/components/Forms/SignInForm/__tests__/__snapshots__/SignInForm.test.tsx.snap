// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInForm /> should create a snapshot for SignInForm 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div>
    <label>
      Username
      <input
        aria-describedby="password-constraints"
        aria-invalid={false}
        autoComplete="username"
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="text"
      />
    </label>
  </div>
  <div>
    <label>
      Password
      <input
        aria-describedby="password-constraints"
        aria-invalid={false}
        autoComplete="current-password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="password"
      />
    </label>
  </div>
  <button
    disabled={true}
    type="submit"
  >
    Submit
  </button>
</form>
`;
